Add Ollama-webui package and service for Mixtral

malteneuss opened the pull request:
## Description of changes

Related to  #273556:
 
- Add [Ollama-WebUI](https://github.com/ollama-webui/ollama-webui) package that mimics ChatGPT's frontend and integrates nicely with existing Ollama package.
- Extend the Ollama service module for NixOS with options for home lab deployments
- Add Ollama-WebUI service module for NixOS

Background:
To make open-source large langue models (LLM) accessible there are projects like [Ollama](https://github.com/NixOS/nixpkgs/issues/ollama.ai) that make it almost trivial to download and run them locally on a consumer computer.
We already have [Ollama in Nixpkgs](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/tools/misc/ollama/default.nix), but that can only be run conveniently in a terminal (and doesn't store previous chats). What's missing is a web UI, e.g. [Ollama-WebUI](https://github.com/ollama-webui/ollama-webui) that mimics ChatGPT's frontend and integrates nicely with Ollama.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

malteneuss committed:
ollama-webui: init at 0.0.0-unstable-2023-12-22

To make open-source large langue models (LLM) accessible there are
projects like Ollama that make it almost trivial to download and
run them locally on a consumer computer. We already have Ollama in
Nixpkgs, but that can only be run conveniently in a terminal
(and doesn't store previous chats). What's missing is a web UI,
e.g. Ollama-WebUI that mimics ChatGPT's frontend and integrates nicely
with Ollama.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-package-static-single-page-nodejs-webapp-ollama-webui/37074/1

malteneuss committed:
nixos/ollama-webui: init service

Make it convenient to setup Ollama-Webui on a server to run
large language models (LLM).

trzpiot commented:
Hi @malteneuss, thanks for packaging Ollama Web UI. 💙 In the meantime, Ollama has been added as a service (but I haven't tested the service yet).[^1] What's missing for bringing the Ollama Web UI to Nixpkgs? 😊 

[^1]: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/services/misc/ollama.nix

mschwaig commented:
As-is this does not build, but I have been running this PR with a few changes to get it to build locally for a few days and it's been working well. I have not looked at what would be required to land it in nixpkgs.
https://github.com/mschwaig/nixpkgs/tree/ollama-mixtral-services

malteneuss commented:
@trzpiot I wanted to add an automatic NixOS test (since the ollama-webui needs to closely follow ollama; otherwise it broke a few times), but i haven't had time to learn to setup this up. Maybe a manual test would suffice for now to get it merged. Thanks for redirecting me to the existing ollama service (someone was faster ;) but i still would like to add some more knobs to make Ollama be deployable to a home lab server).

@mschwaig Thanks for testing it out. I pushed similar fixes to yours into the MR. I will try to test the setup during the weekend. Maybe you could do the same?

mschwaig commented:
I am running this branch since yesterday.

Two things that I noticed were
1. that my existing model files disappeared. Probably this happens because I got `ollama` from the previous version of this branch before, and now I'm getting the one that got merged in the meantime and they store their models in different locations. I don't think that's an issue.
2. that long prompts made the connection between `ollama-web-ui` and `ollama` fail. I did not have enough time to look into this issue. If I have time I will test if the same thing also happens with `oterm` from this branch. I don't think thats an issue with this branch either.

Overall it seems to be running fine.

When you have added the extra config options that you still wanted to add I think you could remove the draft flag from this PR and at that point we can get someone to take a look who feels confident in reviewing the systemd services configs.

happysalada commented on nixos/modules/services/web-apps/ollama-webui.nix:
```diff
@@ -0,0 +1,94 @@
+
+      ollama-webui-package = mkPackageOption pkgs "ollama-webui" { };
+
+      host = mkOption {
```
this host isn't used anywhere, is that intentional ?

happysalada commented on nixos/modules/services/web-apps/ollama-webui.nix:
```diff
@@ -0,0 +1,94 @@
+      in {
+        ExecStart = "${cfg.ollama-webui-package}/bin/ollama-webui --port ${toString cfg.port} ${cors-arg}";
+        DynamicUser = "true";
+        Type = "simple";
```
just for information the default type is simple for services is simple. I'm not sure how I feel about having it written explicitely, if you prefer it that way, feel free to keep it.

happysalada commented on nixos/modules/services/web-apps/ollama-webui.nix:
```diff
@@ -0,0 +1,94 @@
+        Optional: This module is configured to run locally, but can be served from a (home) server,
+        ideally behind a secured reverse-proxy.
+        Look at <https://nixos.wiki/wiki/Nginx> or <https://nixos.wiki/wiki/Caddy>
+        on how to set up a reverse proxy.
```
do you think referencing the caddy file of the original repo would be a good idea ?
https://github.com/ollama-webui/ollama-webui/blob/main/Caddyfile.localhost

happysalada commented on pkgs/tools/misc/ollama-webui/default.nix:
```diff
@@ -0,0 +1,62 @@
+  PUBLIC_API_BASE_URL = "http://localhost:11434/api";
+
+  # The path '/ollama/api' will be redirected to the specified backend URL
+  OLLAMA_API_BASE_URL = PUBLIC_API_BASE_URL;
```
the env var only seems to be set after the build has succeeded
https://github.com/ollama-webui/ollama-webui/blob/main/Dockerfile#L22
does the build fail without ?
if not, this is probably something that should be set in the service, not in the package.

happysalada commented on pkgs/tools/misc/ollama-webui/default.nix:
```diff
@@ -0,0 +1,62 @@
+    mkdir -p $out/lib
+    cp -R ./build/. $out/lib
+
+    mkdir -p $out/bin
```
you don't need those last instructions.
I could be wrong, but if this is a standard node project, you just need to package the build directory
then in the service you can use node directly to just run the build directory
here are some reference of how we did this with lemmy
https://github.com/NixOS/nixpkgs/blob/aa5d1d03b1981f0968e3fdd0cd376c0a06fcaeef/pkgs/servers/web-apps/lemmy/ui.nix#L35
https://github.com/NixOS/nixpkgs/blob/aa5d1d03b1981f0968e3fdd0cd376c0a06fcaeef/nixos/modules/services/web-apps/lemmy.nix#L4

second point and this is entirely optional.
rather than making a complety separate service for ollama-webui, how about including it in the original ollama service.
I don't think it makes sense to have ollama-webui as a standalone for now.
(you can take inspiration in the lemmy module for what we did if you find stuff that you like).

mschwaig replied:
> second point and this is entirely optional.
> rather than making a complety separate service for ollama-webui, how about including it in the original ollama service.
> I don't think it makes sense to have ollama-webui as a standalone for now.
> (you can take inspiration in the lemmy module for what we did if you find stuff that you like).

Since `ollama-webui` is a community project and not officially associated with `ollama` (see the top of [their README.md](https://github.com/ollama-webui/ollama-webui)), I tend to think it does not make sense to have the `ollama-webui` service live under `ollama` as `ollama.ui` or `ollama.web-ui`.

There are a bunch of other frontends that people might want to use with the `ollama` service, such as `oterm`.

happysalada commented:
hey, I've provided a couple of comments, but all in all, this is very nice! Thank you for your contribution. I'm here to help, so let me know if anything!

malteneuss commented:
@happysalada @mschwaig Thanks for the review comments and pushing this topic further. I've been busy with my full-time job and a two year old, and have little to no time to move this forward in the next months. Would your branch be stable and usable enough to be merged and improved in smaller steps?

Kreyren commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -1,42 +1,106 @@
+        OLLAMA_ORIGINS = cfg.cors_origins;
+        # Where to store LLM model files.
+        # Directory is managed by systemd DynamicUser feature, see below.
+        HOME = "%S/ollama";
+        OLLAMA_MODELS = "%S/ollama/models";
```
Isn't it better to use `/var/lib/private/ollama` instead of the `%s` so that it's more functional to work with the variables?

![Uploading image.png…]()

^ Having to hard-code the path as the expansion doesn't seem to work like other services

Kreyren commented on nixos/modules/services/web-apps/ollama-webui.nix:
```diff
@@ -0,0 +1,94 @@
+
+in {
+  options = {
+    services.ollama-webui = {
```
What's the current expected status of this?

Kreyren commented:
> @happysalada @mschwaig Thanks for the review comments and pushing this topic further. I've been busy with my full-time job and a two year old, and have little to no time to move this forward in the next months. Would your branch be stable and usable enough to be merged and improved in smaller steps? -- @malteneuss (https://github.com/NixOS/nixpkgs/pull/275448#issuecomment-2041202792)

Is this still on-going?

* yes -> Please share the current expected status so that i know where to start from as i might have time to finish this MR
* no -> cool cool

mschwaig commented:
@Kreyren

As far as I remember the chat functionality is working on my branch (linked above).
I have not tested that RAG is working yet, which is the feature that made adding all that python necessary.
Some of the python dependencies had to be packaged, some had open PRs which might have been closed in the mean time.

I do not have a lot of time to look into this in the next two weeks.
My next steps would have been to take another look at my branch with the goal of opening a draft PR once I get to it, maybe rebase onto recent master and try to build a newer version of the package, but if someone else wants to pick this up and open a PR in the mean time, go ahead.

I would recommend that you use my branch as a starting point.

malteneuss commented:
Already done with https://github.com/NixOS/nixpkgs/pull/316248
